File: Agora-RTC-With-Voice-Changer-iOS-Objective-C/Agora-RTC-With-Voice-Changer-iOS-Objective-C/MainViewController.m
//
//  MainViewController.m
//  Agora-RTC-With-Voice-Changer-iOS-Objective-C
//
//  Created by ZhangJi on 2018/5/4.
//  Copyright © 2018 ZhangJi. All rights reserved.
//

#import "MainViewController.h"
#import "RoomViewController.h"

@interface MainViewController ()
@property (weak, nonatomic) IBOutlet UIButton *joinButton;
@property (weak, nonatomic) IBOutlet UITextField *channelTextField;

@end

@implementation MainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    NSString *segueId = segue.identifier;
    
    if ([segueId isEqualToString:@"mainToRoom"]) {
        RoomViewController *RoomVC = segue.destinationViewController;
        RoomVC.roomName = self.channelTextField.text;
        RoomVC.clientRole = [sender integerValue];
    }
}

- (void)showRoleSelection {
    UIAlertController *sheet = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    UIAlertAction *broadcaster = [UIAlertAction actionWithTitle:@"Broadcaster" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self joinWithRole:AgoraClientRoleBroadcaster];
    }];
    UIAlertAction *audience = [UIAlertAction actionWithTitle:@"Audience" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self joinWithRole:AgoraClientRoleAudience];
    }];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleDefault handler:nil];
    [sheet addAction:broadcaster];
    [sheet addAction:audience];
    [sheet addAction:cancel];
    [sheet popoverPresentationController].sourceView = self.joinButton;
    [sheet popoverPresentationController].permittedArrowDirections = UIPopoverArrowDirectionUp;
    [self presentViewController:sheet animated:YES completion:nil];
}

- (void)joinWithRole:(AgoraClientRole)role {
    [self performSegueWithIdentifier:@"mainToRoom" sender:@(role)];
}

- (IBAction)doJoinButtonPressed:(id)sender {
    if (self.channelTextField.text.length) {
        [self showRoleSelection];
    }
}

@end

File: Agora-RTC-With-Voice-Changer-iOS/Agora-RTC-With-Voice-Changer-iOS/MainViewController.swift
//
//  MainViewController.swift
//  Agora-RTC-With-Voice-Changer-iOS
//
//  Created by GongYuhua on 2017/4/7.
//  Copyright © 2017年 Agora. All rights reserved.
//

import UIKit

class MainViewController: UIViewController {
    
    @IBOutlet weak var roomNameTextField: UITextField!
    @IBOutlet weak var joinButton: UIButton!
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let segueId = segue.identifier , segueId == "mainToRoom" else {
            return
        }
        
        let roomVC = segue.destination as! RoomViewController
        roomVC.roomName = roomNameTextField.text
        if let value = sender as? NSNumber, let role = AgoraClientRole(rawValue: value.intValue) {
            roomVC.clientRole = role
        }
        roomVC.delegate = self
    }
    
    @IBAction func doJoinPressed(_ sender: UIButton) {
        enterRoom()
    }
}

private extension MainViewController {
    func enterRoom() {
        guard let roomName = roomNameTextField.text , !roomName.isEmpty else {
            return
        }
        
        let sheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let broadcaster = UIAlertAction(title: "Broadcaster", style: .default) { [unowned self] _ in
            self.enterRoom(withRole: .broadcaster)
        }
        let audience = UIAlertAction(title: "Audience", style: .default) { [unowned self] _ in
            self.enterRoom(withRole: .audience)
        }
        let cancel = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        sheet.addAction(broadcaster)
        sheet.addAction(audience)
        sheet.addAction(cancel)
        sheet.popoverPresentationController?.sourceView = joinButton
        sheet.popoverPresentationController?.permittedArrowDirections = .up
        present(sheet, animated: true, completion: nil)
    }
    
    func enterRoom(withRole role: AgoraClientRole) {
        performSegue(withIdentifier: "mainToRoom", sender: NSNumber(value: role.rawValue))
    }
}

extension MainViewController: RoomVCDelegate {
    func roomVCNeedClose(_ roomVC: RoomViewController) {
        dismiss(animated: true, completion: nil)
    }
}

extension MainViewController: UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        enterRoom()
        return true
    }
}
